packages/boxplot/src/compute/stratification.ts

```diff
@@ -0,0 +1,101 @@
+import { BoxPlotDatum, BoxPlotCommonProps, BoxPlotSummary } from '../types'
+import { defaultProps } from '../props'
+
+/** stratify an array of raw data objects into an array of arrays;
+ * each array will create one box plot */
+export const stratifyData = <RawDatum extends BoxPlotDatum>({
+    data,
+    groups = defaultProps.groups,
+    getGroup,
+    subGroups = defaultProps.subGroups,
+    getSubGroup,
+}: {
+    data: RawDatum[]
+    groups?: BoxPlotCommonProps<RawDatum>['groups']
+    getGroup: ((datum: RawDatum) => string) | null
+    subGroups?: BoxPlotCommonProps<RawDatum>['subGroups']
+    getSubGroup: ((datum: RawDatum) => string) | null
+}) => {
+    const groupsMap = {} as BoxPlotDatum
+    if (groups) {
+        groups.map((g, i) => (groupsMap[g] = i))
+    }
+    const subGroupsMap = {} as BoxPlotDatum
+    if (subGroups) {
+        subGroups.map((sg, i) => (subGroupsMap[sg] = i))
+    }
+
+    const nGroups = Math.max(1, groups ? groups.length : 1)
+    const nSubGroups = Math.max(1, subGroups ? subGroups.length : 1)
+    const n = nGroups * nSubGroups
+    const result = Array.from(Array(n), () => [] as RawDatum[])
+    data.map((d: RawDatum) => {
+        const groupIndex = getGroup ? Number(groupsMap[getGroup(d)]) : 0
+        const subGroupIndex = getSubGroup ? Number(subGroupsMap[getSubGroup(d)] ?? 0) : 0
+        const index = groupIndex * nSubGroups + subGroupIndex
+        if (index >= 0) {
+            result[index].push(d)
+        }
+        return null
+    })
+    return result
+}
+
+const getQuantile = (values: number[], quantile = 0.5) => {
+    const realIndex = (values.length - 1) * Math.max(0, Math.min(1, quantile))
+    const intIndex = Math.floor(realIndex)
+    if (realIndex === intIndex) return values[intIndex]
+    const v1 = values[intIndex],
+        v2 = values[intIndex + 1]
+    return v1 + (v2 - v1) * (realIndex - intIndex)
+}
+
+const getMean = (values: number[]) => {
+    const sum = values.reduce((acc, x) => acc + x, 0)
+    return sum / values.length
+}
+
+export const summarizeDistributions = <RawDatum extends BoxPlotDatum>({
+    data,
+    getValue,
+    groups,
+    subGroups,
+    groupIndex,
+    subGroupIndex,
+    quantiles,
+}: {
+    data: RawDatum[]
+    getValue: (datum: RawDatum) => unknown
+    groups: string[] | null
+    subGroups: string[] | null
+    groupIndex: number
+    subGroupIndex: number
+    quantiles: number[]
+}) => {
+    // accept precomputed summary representations if they have all the required keys
+    const preComputedKeys = ['values', 'extrema', 'mean', 'quantiles', 'group', 'subGroup', 'n']
+    if (data.length === 1) {
+        const isPrecomputed = preComputedKeys.reduce((acc, k) => acc && k in data[0], true)
+        if (isPrecomputed) {
+            return {
+                groupIndex: groupIndex,
+                subGroupIndex: subGroupIndex,
+                ...data[0],
+            } as unknown as BoxPlotSummary
```

Comment:
I think we could avoid the casting here by having using a [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates).

Reply:
Thank you. I learned several things through this review; this is one of them.

I read through the doc and tried to implement something along those lines. Could you check if this is what you meant? Maybe there is a less verbose way to write the function that provides the type predicate? I can keep editing to improve.

Reply:
Yes, that's what I meant.